### PR TITLE
feat: add aws batch compute and job queue

### PR DESCRIPTION
**Pull request by sachinmalanki**
### Description of your changes

Adding AWS Batch Compute Environment(aws_batch_compute_environment) and Job Queue(aws_batch_job_queue)

Fixes #

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested
Manually tested as per the guide
Batch compute environment
image

Job Queue
image

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by sachinmalanki** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by sachinmalanki** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by sachinmalanki** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by sachinmalanki** [comment]
For the test to fails because of hardcoding securityGroupIds: and subnets: values. Can I know how to fix this? The computeenvironment crd does not allow subnet and securitygroupid reference.

**Comment by mergenci** [comment]
@sachinmalanki, here's an example configuration: https://github.com/crossplane-contrib/provider-upjet-aws/blob/v1.14.0/config/apprunner/config.go#L14-L23

**Comment by sachinmalanki** [comment]
Thanks for the help @mergenci, I have updated the MR with reference and selector.

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/computeenvironment.yaml"

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml,examples/batch/v1beta1/computeenvironment.yaml"

**Comment by sachinmalanki** [comment]
Looks like computeEnvironment resource fails to delete because of `statusReason: 'CLIENT_ERROR - User: batch.amazonaws.com is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::153891904029:role/aws-batch-service-role (Service: AWSSecurityTokenService; Status Code: 403; Error Code: AccessDenied; Request ID: 3006240f-9511-4141-928e-1dd0258ebb24; Proxy: null)' 
I believe this is because the service role is deleted before the MR is deleted. 
Similar issue - https://github.com/hashicorp/terraform-provider-aws/issues/8549
Any idea, how to add dependency so computeEnvironment is deleted before the serviceRole?

**Comment by jeanduplessis** [comment]
/test-examples="examples/batch/v1beta1/jobqueue.yaml,examples/batch/v1beta1/computeenvironment.yaml"

**Comment by sachinmalanki** [comment]
Hi, I'm stuck again. Could anyone please help?
trying to get aws_batch_compute_environment resource added but it fails on uptest.
```
apiVersion: batch.aws.upbound.io/v1beta1
kind: ComputeEnvironment
metadata:
  annotations:
    meta.upbound.io/example-id: batch/v1beta1/computeenvironment
  labels:
    testing.upbound.io/example-name: sample_ce
  name: sample
spec:
  forProvider:
    eksConfiguration: {}
    computeResources:
      instanceType:
      - c4.large
      maxVcpus: 16
      instanceRoleSelector:
        matchLabels:
          testing.upbound.io/example-name: ecs_instance_role
      placementGroupSelector:
        matchLabels:
          testing.upbound.io/example-name: sample_pg
      securityGroupSelector:
        matchLabels:
          testing.upbound.io/example-name: sample_sg
      subnetSelector:
        matchLabels:
          testing.upbound.io/example-name: sample_subnet
      type: EC2
    region: eu-west-1
    serviceRoleSelector:
      matchLabels:
        testing.upbound.io/example-name: aws_batch_service_role
    type: MANAGED
```

But get the following error which fails to update the external resource -
`update failed: async update failed: refuse to update the external resource because the following update requires replacing it: cannot change the value of the argument "eks_configuration.#" from "0" to "1"`
And when  eksConfiguration  is remove, getting the following error
`Warning  CannotUpdateExternalResource  9m37s (x2 over 10m)  managed/batch.aws.upbound.io/v1beta1, kind=computeenvironment  (combined from similar events): async update failed: failed to set observation: v1beta1.ComputeEnvironmentObservation.EksConfiguration: readObjectStart: expect { or n, but found [, error found in #10 byte of ...|uration":[],"service|..., bigger context ...|e_environment_name":"sample","eks_configuration":[],"service_role":"arn:aws:iam::811861054499:role/a|...`
Any inputs on how to resolve this?

File: config/externalname.go [hunk 1]
```diff
@@ -29,6 +29,9 @@ var TerraformPluginFrameworkExternalNameConfigs = map[string]config.ExternalName
 	// terraform-plugin-framework
 	"aws_appconfig_environment": appConfigEnvironment(),
 
+	// AWS Batch job queue can be imported using the name
+	"aws_batch_job_queue": config.NameAsIdentifier,
```
Comment (turkenf) [inline config/externalname.go:33]: In the terraform document, I see that the resource will be imported with arn in the [import](https://registry.terraform.io/providers/hashicorp/aws/5.68.0/docs/resources/batch_job_queue#import) section. Therefore, in order to use the [NameAsIdentifier](https://github.com/crossplane/upjet/blob/main/docs/adding-new-resource.md#case-1-name-as-identifier) configuration, the arn of the resource and the name parameters must be the same. As I see they are not same so here we can use [IdentifierFromProvider](https://github.com/crossplane/upjet/blob/main/docs/adding-new-resource.md#case-3-random-identifier-from-provider) or TemplatedStringAsIdentifier, please try below:

```suggestion
	"aws_batch_job_queue": config.TemplatedStringAsIdentifier("name", "arn:aws:batch:{{ .setup.configuration.region }}:{{ .client_metadata.account_id }}:job-queue/{{ .external_name }}"),
```

File: examples/batch/v1beta1/computeenvironment.yaml [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+spec:
+  forProvider:
+    region: us-west-1
+    strategy: cluster
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: Policy
+metadata:
+  annotations:
+    meta.upbound.io/example-id: iam/v1beta1/role
```
Comment (turkenf) [inline examples/batch/v1beta1/computeenvironment.yaml:150]:
```suggestion
    meta.upbound.io/example-id: batch/v1beta1/computeenvironment
```

File: examples/batch/v1beta1/computeenvironment.yaml [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                  }
+              }
+          ]
+      }
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: Policy
+metadata:
+  annotations:
+    meta.upbound.io/example-id: iam/v1beta1/role
```
Comment (turkenf) [inline examples/batch/v1beta1/computeenvironment.yaml:240]:
```suggestion
    meta.upbound.io/example-id: batch/v1beta1/computeenvironment
```

File: examples/batch/v1beta1/jobqueue.yaml [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        order: 1
+    priority: 1
+    region: us-west-1
+    state: ENABLED
+---
+apiVersion: batch.aws.upbound.io/v1beta1
+kind: ComputeEnvironment
+metadata:
+  annotations:
+    meta.upbound.io/example-id: batch/v1beta1/computeenvironment
```
Comment (turkenf) [inline examples/batch/v1beta1/jobqueue.yaml:24]:
```suggestion
    meta.upbound.io/example-id: batch/v1beta1/jobqueue
```

And please fix the above example-ids.

File: package/crossplane.yaml.tmpl [hunk 5]
```diff
@@ -27,7 +27,7 @@ metadata:
       If you encounter an issue please reach out on support@upbound.io email
       address. This is a subpackage for the {{ .Service }} API group.
     friendly-name.meta.crossplane.io: Provider AWS ({{ .Service }})
-    auth.upbound.io/group: {{ .ProviderName }}.upbound.io
+    auth.upbound.io/group: {{ .ProviderAuthGroup }}.upbound.io
```
Comment (turkenf) [inline package/crossplane.yaml.tmpl:30]: +1

**Review by turkenf — COMMENTED** [review]
Thank you for your effort @sachinmalanki, I left a few comments for you to consider.

**Comment by sachinmalanki** [comment]
Hi @turkenf. Thanks for the suggestions, I have rebased and made the changes. Could you please review ?
When I was testing this locally, the service_role gets deleted first and leaves the computeenvironment in a INVALID hanging state. Any idea how this could be fixed? This could cause the uptest to fail.

**Comment by turkenf** [comment]
> When I was testing this locally, the service_role gets deleted first and leaves the computeenvironment in a INVALID hanging state. Any idea how this could be fixed? This could cause the uptest to fail.

@sachinmalanki, we have [hooks](https://github.com/crossplane/uptest?tab=readme-ov-file#hooks) for this kind of cases. We can use pre-delete-hook. Here is an example: [yaml](https://github.com/crossplane-contrib/provider-upjet-aws/blob/be713161577c2277c3ae539eda3f18015e7f51df/examples/iot/v1beta1/topicruledestination.yaml#L36) and [sh](https://github.com/crossplane-contrib/provider-upjet-aws/blob/main/examples/iot/v1beta1/testhooks/delete-topicruledestination.sh) files.

Also, I see 47 commits in the PR's commit history. Please fork your PR from the latest main and squash your commits.

File: config/batch/config.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+				TerraformName:     "aws_subnet",
+				RefFieldName:      "SubnetRefs",
+				SelectorFieldName: "SubnetSelector",
+			},
+			"compute_resources.security_group_ids": config.Reference{
+				TerraformName:     "aws_security_group",
+				RefFieldName:      "SecurityGroupRefs",
+				SelectorFieldName: "SecurityGroupSelector",
+			},
+		}
```
Comment (turkenf) [inline config/batch/config.go:21]:
```suggestion
		r.References["compute_resources.subnets"] = config.Reference{
			TerraformName:     "aws_subnet",
			RefFieldName:      "SubnetRefs",
			SelectorFieldName: "SubnetSelector",
		}
		r.References["compute_resources.security_group_ids"] = config.Reference{
			TerraformName:     "aws_security_group",
			RefFieldName:      "SecurityGroupRefs",
			SelectorFieldName: "SecurityGroupSelector",
		}
```

**Comment by sachinmalanki** [comment]
I have raised another PR - https://github.com/crossplane-contrib/provider-upjet-aws/pull/1542 with all the changes suggested.

**Comment by turkenf** [comment]
Hi @sachinmalanki, which PR will we continue with?

**Comment by sachinmalanki** [comment]
Lets continue with https://github.com/crossplane-contrib/provider-upjet-aws/pull/1542. let me close this PR

**Pull request closed without merging**